### PR TITLE
Add sanity server/client version check

### PR DESCRIPTION
**Pull request by Lnaden**
Simple check to ensure the client is at least equal to the server
from a minor revision level.

Open to debate on if this is the correct check.

## Status
- [ ] Changelog updated
- [ ] Ready to go

### CODE REVIEW

File: qcfractal/interface/client.py [hunk 1]
```diff
@@ -79,6 +80,18 @@ def __init__(self,
...
         self.server_name = self.server_info["name"]
 
+        def version_list(version):
+            re.sub(r"\+.*$", "", version.strip("v"))
+            return [int(x) for x in version.split(".")]
+
+        from . import __version__  # Import here to avoid circular import from __init__
+        server_version = version_list(self.server_info["version"])
+        client_version = version_list(__version__)
+        assert client_version[:-1] >= server_version[:-1], (f"Sever's' Fractal version of {server_version} appears to "
```
Comment (dgasmith) [inline qcfractal/interface/client.py]: To be save and explicit I would do `[:2]`. Typos is "Sever".

Comment (dgasmith) [inline qcfractal/interface/client.py]: I think for now major/minor must match. Backwards compute is not guaranteed either.

File: qcfractal/interface/client.py [hunk 2]
```diff
@@ -79,6 +80,18 @@ def __init__(self,
 
         self.server_name = self.server_info["name"]
 
+        def version_list(version):
+            re.sub(r"\+.*$", "", version.strip("v"))
```
Comment (dgasmith) [inline qcfractal/interface/client.py]: Does `re` update in place?

Comment (Lnaden) [reply qcfractal/interface/client.py]: No, typo on my part

File: qcfractal/interface/client.py (hunk 1 above)
Comment (Lnaden) [reply qcfractal/interface/client.py]: So right now the server is on 0.5.4 but the developmental stuff we are doing is on 0.6.0, if we make them match, we will stop being able to develop or run production at this moment unless we downgrade. 

I also don't think this PR should be merged in until we have a set upgrade path/procedure in place for the server.

**Comment by dgasmith** [comment]
Link to #191.

**Comment by dgasmith** [comment]
For now, let's have a simple solution where the FractalServer has a minimum and maximum Client version it will connect to. When the client opens the connection to a server it will pull this data from `/information` (which will likely never change) and will make sure the client version is within the bracket. 

This allows uncoupling of the client and rightly puts all onus on the server which is much more controllable. The only issue is transitionary periods and testing. We currently need to pin to `0.7.x`, but that will break testing as the version is not out yet. I will think about that solution, might be possible to do fill git revision values as well.

File: qcfractal/interface/client.py (hunk 2 above)
Comment (loriab) [reply qcfractal/interface/client.py]: better to run `safe_version` on everything?

Comment (dgasmith) [reply qcfractal/interface/client.py]: We control both sides... I think we better convert this to a tuple and ship it that way from the server and be done with it.

**Comment by Lnaden** [comment]
This should be good to go. I'd like to add a test to remind us to update the min/max versions, but Im not sure how to do that since the release is what sets the version numbers.

**Comment by Lnaden** [comment]
I'm not quite sure how to work around the failing tests here. They fail because they are trying to connect to the central, current MolSSI server which doesn't present this information which is now checked for.

**Comment by dgasmith** [comment]
Please rebase the PR, everything should work again.

File: qcfractal/interface/client.py [hunk 3]
```diff
@@ -82,6 +83,27 @@ def __init__(self,
...
+        from . import __version__  # Import here to avoid circular import from __init__
+        try:
+            server_version_min_client = version_list(self.server_info["client_lower_version_limit"])
+            server_version_max_client = version_list(self.server_info["client_upper_version_limit"])
+        except KeyError:
+            raise IOError(f"The Server at {self.address}, version {self.server_info['version']} does not report what "
+                          f"Client versions it accepts! It can be almost asserted your Client is too new for the Server "
+                          f"you are connecting to.")
+        client_version = version_list(__version__)
+        assert server_version_min_client <= client_version <= server_version_max_client, (
```
Comment (dgasmith) [inline qcfractal/interface/client.py]: We should raise an actual error here rather than an assert.

File: qcfractal/interface/client.py [hunk 4]
```diff
@@ -82,6 +83,27 @@ def __init__(self,
...
+                f"This Client of version {client_version} does not fall within the Server's allowed Client versions "
+                f"of [{server_version_min_client}, {server_version_max_client}]. Server address: {self.address}"
```
Comment (dgasmith) [inline qcfractal/interface/client.py]: Perhaps note that it is likely that they need to upgrade their portal version with `pip install qcportal` or `conda install qcportal -c conda-forge.`


These kinds of error messages both here and above should tell a user what to do to fix the issue not just tell them there is an issue.

File: qcfractal/interface/client.py [hunk 5]
```diff
@@ -82,6 +83,27 @@ def __init__(self,
 
         self.server_name = self.server_info["name"]
 
+        def version_list(version):
```
Comment (dgasmith) [inline qcfractal/interface/client.py]: Could we move this above the class with a `_version_list` name? We might want to move this to a util eventually.

File: qcfractal/interface/client.py (hunk 3 above)
Comment (Lnaden) [reply qcfractal/interface/client.py]: It will raise an actual error if the assert fails, and gives the error string which is the 2nd arg to `assert`

Comment (dgasmith) [reply qcfractal/interface/client.py]: Yes, but why use an assert here over a canonical raise exception? If you truly want an assertion error you can `raise AssertionError`.

**Pull request merged**